### ISSUE TITLE
Demo link?

### ISSUE THREAD

**Issue by oyeanuj** [blakeembrey/node-scrappy#13]
Hi @blakeembrey! The project looks great and I'm looking forward to playing with it more. Given the expansive nature of the data you fetch via this project, I was wondering if you plan to have a demo link where folks could play around with urls to see how the library works? 

The project seems to have a lot of cool things to show off :)

**Comment by blakeembrey** [comment]
A demo would be great, but it'd have to be hosted and node-based since it doesn't work so well in browsers (in theory it could, but not currently my focus).

**Comment by MiniGod** [comment]
You could always use runkit. https://runkit.com/npm/scrappy

In fact, I found this module so useful that I put up on now.sh a weeks ago: https://metadata.now.sh/

Example: https://metadata.now.sh/https://medium.com/slack-developer-blog/everything-you-ever-wanted-to-know-about-unfurling-but-were-afraid-to-ask-or-how-to-make-your-e64b4bb9254

**Comment by blakeembrey** [comment]
@MiniGod That's awesome, cheers! If you wanted, that can be added to the README. I just tested runkit and it's erroring on a module being missing, I'll have to retry it when I release the next version (which the result is looking a lot different to the last release). I really should have thought of runkit, I did do some testing of different solutions initially with tonicdev.

**Comment by MiniGod** [comment]
As far as I'm concerned you can use it, but I provide no SLA ;) ... but runkit is probably the way to go?
Concerning the missing module error on runkit, it looks like scrappy only works with npm@3 or newer. But it looks like you fixed that error back in July (13th), but you have not released a new version to npm since a few days before that (July 8th)?
I bet if you get some time to publish a new version to npm then runkit will work?

**Comment by blakeembrey** [comment]
I agree, runkit would be nice. I just released a new version and runkit is still giving me missing module errors - it just doesn't like the dependencies I use. I believe this one is coming from `jsonld`, so if it's not fixed I might have to open an issue against it to try and remove the dependency (it looks pretty unnecessary anyway).

**Comment by MiniGod** [comment]
@blakeembrey scrappy works on runkit now!
Here is the example code from the readme running on runkit: https://runkit.com/minigod/589f242c8cec31001454e3d2

**Comment by oyeanuj** [comment]
@MiniGod For your micro-service, do you have a link to the repo?

@blakeembrey I was thinking something similar, or like https://github.com/mxstbr/micro-open-graph but with `node-scrappy` would be ideal.  Do you think that would be beneficial?

**Comment by blakeembrey** [comment]
@oyeanuj When it's a free `now` plan, you can actually visit the source code: https://metadata.now.sh/_src/?f=index.js.

Can you give more information on what you wanted to do? I might have lost the context of the thread and I'll add the Runkit demo link to the README now 😄 

You can definitely do whatever you need to with this library. Personally, I had built it to extract metadata for in-browser link previews. I do still want to expand the previews to become more and more interactive - for instance, I want to give previews on a product with payment link, map with directions links and more. I'd love to see what everyone else is using it for, and I intend to turn it into a service in the near future (turns out making a product out of an OSS library is tricky - takes so much time to set up the infra for things like billing and logging part-time).

**Comment by oyeanuj** [comment]
@blakeembrey Thanks for the instant response :)  I have a very similar usecase to yours except that I am using Rails on the backend. So, for something extensible, I wanted to run from the browser or a microservice like the one above.

In my case as well, I want to do link previews and incase it is products or maps, do smarter link previews. Ideally, I imagine it would be an XHR call which gets that information. Maybe I can build out a microservice with those different kind of information, similar to `metadata` or `micro-open-graph` links above, or add to them if they are interested.

I am curious - do you get the information on on server-side with page load or through a micro-service or XHR call in your usage? Where do you see it - publicly available?

And finally, I totally understand OSS is a lot of investment - thank you already for putting out this work. It is much appreciated! :)

**Comment by blakeembrey** [comment]
> do you get the information on on server-side with page load or through a micro-service or XHR call in your usage?

I just had a queue set up on the back-end that would process links into snippets and cache for 24 hours. You wouldn't want to do it on page load because your service would become dependent on an external URL processing time (which could take a **long** time to respond). I did also have an XHR endpoint that was used for inline previews, but I just made that long poll since, depending on the queue, you may not know when the link preview is available and didn't want to hold the connection open for longer than needed.

> Where do you see it - publicly available?

Sorry, all that work was for a different company and private. But, it was all the same snippet information I've made publicly available here - nothing extra (except slapping on a nice template and video previews, etc). I never got to requiring the extra information, even though it would be possible, since I didn't have enough time to build it that before I had to build other features.

**Comment by oyeanuj** [comment]
Thanks for the detailed answer! One last question (I know this is going off-topic with the issue) - but I was wondering how you compare it to Ian's metascraper - advantages/disadvantages, if any?

**Comment by blakeembrey** [comment]
I talked with Ian a lot when I was first launching and he took a look at this library too for some parts (e.g. he adopted popsicle for requests and I pointed out that he needed a cookie store on node, etc). We both decided that our use-cases differed enough from each other to keep both libraries separate, but perhaps that could change as the architecture from Scrappy is now pluggable. 

Basically, he wanted a very specific use-case, which was extracting article information with a no-matter-what approach (it looks for CSS selectors through a waterfall approach) while Scrappy was always intended to be 100% accurate to extract extremely rich information (at the expense that I don't try to magically infer anything based on class-names). I also designed my library from the get-go with the extreme limitation of being 100% streaming, while metascraper buffers the data and works synchronously. Not a big deal there for article information, but I'm getting information from _everything_ - which may include a 100MB video, for instance, and I would want to drop that request stream after I've read enough header bytes to infer the metadata.
